### PR TITLE
FOGL-1445 Move categories into a hierarchy

### PR DESCRIPTION
**Pull request by MarkRiddoch**

### CODE REVIEW

**Comment by amarendra-dianomic** [comment]
@MarkRiddoch the errors as I reported before appears to have been fixed. But @ashish-jabble mentioned, the code errors need to be fixed also. Also, test fails to detect that cfg_manager.create_child_category() is not part of this PR + coroutines are called without "await", so we may need to refine tests also but this can be taken up later separately.

**Comment by amarendra-dianomic** [comment]
@MarkRiddoch FOGL-1445 branch also needs to be rebased as there has been major change in dependencies.

File: python/foglamp/services/core/server.py [hunk 1]
```diff
@@ -492,6 +492,27 @@ def _write_pid(cls, api_address, api_port):
...
 
+    @classmethod
+    async def _config_parents(cls):
+        # Create the parent category for all general configuration categories
+        try:
+            await cls._configuration_manager.create_category("General", {}, 'General', True)
+            await cls._configuration_manager.create_child_category("General", ["service","rest_api"])
+        except KeyError:
+            message = self._MESSAGES_LIST['e000004'].format(self._name)
+            _LOGGER.error(message)
```
Comment (ashish-jabble) [inline python/foglamp/services/core/server.py]: use `_logger` here

File: python/foglamp/services/core/server.py [hunk 2]
```diff
@@ -492,6 +492,27 @@ def _write_pid(cls, api_address, api_port):
...
             sys.exit(1)
 
+    @classmethod
+    async def _config_parents(cls):
+        # Create the parent category for all general configuration categories
+        try:
+            await cls._configuration_manager.create_category("General", {}, 'General', True)
+            await cls._configuration_manager.create_child_category("General", ["service","rest_api"])
+        except KeyError:
+            message = self._MESSAGES_LIST['e000004'].format(self._name)
```
Comment (ashish-jabble) [inline python/foglamp/services/core/server.py]: Please fix attribute reference for _name and _MESSAGES_LIST

File: python/foglamp/services/core/server.py [hunk 3]
```diff
@@ -492,6 +492,27 @@ def _write_pid(cls, api_address, api_port):
...
+            _LOGGER.error(message)
+            raise
+
+        # Create the parent category for all advanced configuration categories
+        try:
+            await cls._configuration_manager.create_category("Advanced", {}, 'Advanced', True)
+            await cls._configuration_manager.create_child_category("Advanced", ["SMNTR", "SCHEDULER"])
+        except KeyError:
+            message = self._MESSAGES_LIST['e000004'].format(self._name)
```
Comment (ashish-jabble) [inline python/foglamp/services/core/server.py]: fix attribute reference for _name and _MESSAGES_LIST

File: python/foglamp/services/core/server.py [hunk 4]
```diff
@@ -492,6 +492,27 @@ def _write_pid(cls, api_address, api_port):
...
+            _LOGGER.error(message)
```
Comment (ashish-jabble) [inline python/foglamp/services/core/server.py]: use _logger

File: python/foglamp/tasks/north/sending_process.py [hunk 5]
```diff
@@ -1076,6 +1077,17 @@ def _fetch_configuration(self, cat_name=None, cat_desc=None, cat_config=None, ca
...
             self._core_task_management_client.create_configuration_category(config_payload)
             _config_from_manager = self._core_task_management_client.get_configuration_category(category_name=cat_name)
+
+            # Create the parent category for all north services
+            try:
+                parent_payload = json.dumps({"key": "North", "description":"North tasks","value":{},
+                    "children": [cat_name], "keep_original_items":True})
+                self._core_task_management_client.create_configuration_category(parent_payload)
+            except KeyError:
+                message = self._MESSAGES_LIST['e000030'].format(self._name)
```
Comment (ashish-jabble) [inline python/foglamp/tasks/north/sending_process.py]: Fix attribute reference for _name and  _MESSAGES_LIST

**Comment by aksinha-nerdapplabs** [comment]
@MarkRiddoch "Advanced" and "Processes" parent categories are not listed in [FOGL-1445](https://scaledb.atlassian.net/browse/FOGL-1445) description. Do we need to update description in the ticket or modify categories here?

**Pull request merged**